lib/rubocop/rspec/expect_offense.rb

```diff
@@ -128,13 +128,16 @@ def expect_correction(correction, loop: false)
 
           break corrected_source unless loop
           break corrected_source if cop.corrections.empty?
+          break corrected_source if corrected_source == @processed_source.buffer.source
 
           if iteration > RuboCop::Runner::MAX_ITERATIONS
             raise RuboCop::Runner::InfiniteCorrectionLoop.new(@processed_source.path, [])
           end
 
           # Prepare for next loop
           cop.instance_variable_set(:@corrections, [])
+          # Cache invalidation. This is bad!
+          cop.instance_variable_set(:@token_table, nil)
```

Comment:
The mixin `SurroundingSpace` mixin should define `investigate` (`on_before_walk` after my PR lands) to clear its instance variables. If you prefer you can leave things like this since I'll have to tweak it in one case or the other...

Reply:
If `SurroundingSpace` implements `investigate`, we might want to also implement `investigate` in `Cop` to reset the `@corrections` array. But then all existing `investigate` methods should also call `super`…

Knowing that you have some changes coming, I’d prefer to just leave the changes as they are, even with the “This is bad!” comment. WDYT?

Reply:
> we might want to also implement investigate in Cop to reset the @corrections array

My refactor indeed does that automatically, but not in an official callback that could be mistakenly overriden. I will add in the code documentation that it is recommended to call super anyways though.

> I’d prefer to just leave the changes as they are

Sure, as I wrote either way is fine by me.